designs/2019-config-simplification/README.md

```diff
@@ -0,0 +1,950 @@
+
+```js
+module.exports = {
+    name: "eslint-config-example",
+
+    // other info here
+};
+```
+
+It's recommended that the shareable config provide a unique name for each config that is exported.
+
+#### Disambiguating Shareable Configs With Common Dependencies
+
+Today, shareable configs that depend on plugin rules must specify the plugin as a peer dependency and then either provide a script to install those dependencies or ask the user to install them manually.
+
+With this design, shareable configs can specify plugins as direct dependencies that will be automatically installed with the shareable config, improving the user experience of complex shareable configs. This means it's possible for multiple shareable configs to depend on the same plugin and, in theory, depend on different versions of the same plugin. In general, npm will handle this directly, installing the correct plugin version at the correct level for each shareable config to `require()`. For example, suppose there are two shareable configs, `eslint-config-a` and `eslint-config-b` that both rely on the `eslint-plugin-example` plugin, but the former relies on 1.0.0 and the latter relies on 2.0.0. npm will install those plugins like this:
+
+```
+your-project
+├── eslint.config.js
+└── node_modules
+  ├── eslint
+  ├── eslint-config-a
+  | └── node_modules
+  |   └── eslint-plugin-example@1.0.0
+  └── eslint-config-b
+    └── node_modules
+      └── eslint-plugin-example@2.0.0
+```
+
+The problem comes when the shareable configs try to use the default namespace of `eslint-plugin-example` for its rules, such as:
+
+```js
+const example = require("eslint-plugin-example");
+
+module.exports = {
+    plugins: {
+        example
+    }
+};
+```
+
+If both shareable configs do this, and the user tries to use both shareable configs, an error will be thrown when the configs are normalized because the plugin namespace `example` can only be assigned once.
+
+To work around this problem, the end user can create a separate namespace for the same plugin so that it doesn't conflict with an existing plugin namespace from a shareable config. For example, suppose you want to use `eslint-config-first`, and that has an `example` plugin namespace defined. You'd also like to use `eslint-config-second`, which also has an `example` plugin namespace defined. Trying to use both shareable configs will throw an error because a plugin namespace cannot be defined twice. You can still use both shareable configs by creating a new config from `eslint-config-second` that uses a different namespace. For example:
+
+```js
+// get the config you want to extend
+const configToExtend = require("eslint-config-second");
+
+// create a new copy (NOTE: probably best to do this with @eslint/config somehow)
+const compatConfig = Object.create(configToExtend);
+compatConfig.plugins["compat::example"] = require("eslint-plugin-example");
+delete compatConfig.plugins.example;
```

Comment:
I'm not sure if this is valid because shareable configs may be an array or a function. Is there a more general way?